### PR TITLE
feat: Add mongodb as a default database for application

### PR DESCRIPTION
**Pull request by mimotej**
# Related issues
Resolves https://github.com/OqixDevs/OqixTS/issues/60

# Type of pull request
- [x] New feature
- [ ] Bug fix
- [ ] Refactoring
- [ ] Feature update
- [ ] Other

# Describe your PR in detail
This PR adds mongodb database as a default database for the application. It utilizes mongodb atlas [shared tier](https://www.mongodb.com/pricing) since it is fairly generous, later with the application would grow too large we might want to reconsider alternatives...
It uses [Prisma](https://www.prisma.io/) as ORM 
Also adds new states for status -> removed and banned, where removed users can reverify but banned users can't.

### CODE REVIEW

**Comment by mimotej** [comment]
Ok I think it is ready for the review.
/cc @xvanick1

File: src/commands/verify.ts [hunk 1]
```diff
@@ -50,39 +42,61 @@ export async function execute(interaction: ChatInputCommandInteraction) {
...
+    } catch (err) {
+        console.log(`Database error: ${err}`);
+        return interaction.editReply({
+            content: 'Verification failed! Contact admin.',
+        });
+    }
+    if (user.length != 0) {
+        return interaction.editReply({
+            content:
+                'User already verified! Contact admin if you need to verify again.',
```
Comment (xvanick1) [inline src/commands/verify.ts:66]: Shouldn't there be a possibility for user to reverify for some reason (new account/change of spec)? But in that case, we would need to remove the previous verification where this thesis was used.

File: src/model/schema.prisma [hunk 2]
```diff
@@ -0,0 +1,15 @@
+generator client {
+  provider = "prisma-client-js"
+}
+
+datasource database {
+  provider = "mongodb"
+  url      = env("DATABASE_CONNECTION_STRING")
+}
+
+model Users {
```
Comment (xvanick1) [inline src/model/schema.prisma:10]: + date of verification

File: tests/verify.test.ts [hunk 3]
```diff
@@ -16,23 +16,23 @@ const dict = {
 };
 let nameUser = ' Josef Novak';
 
+const databaseUser = {
+    id: '123',
+    discordId: '123',
+    idThesis: '2223121',
+    status: 'verified',
+};
+
```
Comment (xvanick1) [inline tests/verify.test.ts:25]: huh?

File: src/model/schema.prisma [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+  id       String  @id @default(auto()) @map("_id") @database.ObjectId
+  discordId      String  @map("id")
+  idThesis String?
+  status   String
+}
```
Comment (xvanick1) [inline src/model/schema.prisma:15]: plus date of verification

**Review by xvanick1 — COMMENTED** [review]
Suggestion for the possibility of re-verifying with the requirement of deleting old verification assigned to the thesis. Request to add the date of verification to the user object.

**Comment by mimotej** [comment]
> Suggestion for the possibility of re-verifying with the requirement of deleting old verification assigned to the thesis. Request to add the date of verification to the user object.

Reverify option is little bit complicated since we need to define when is user able to reverify - For example there is big difference when the user is kicked from the server and when he leaves on its own. Is only situation when he can reverify when he leaves on its own or do we want to have option to "lift the ban". If the second option is true, then I would suggest leaving it for now and making it separate feature.

As for the the date of verification. Do you have something in mind how we would use it? Because first of all I would like to keep amount of data we store to minimum (mainly to protect students) and we also have currently quite a few people who are already verfied therefore simplest thing would be to add same date to each one, and I do not find that too useful :thinking: .

**Comment by xvanick1** [comment]
> > Suggestion for the possibility of re-verifying with the requirement of deleting old verification assigned to the thesis. Request to add the date of verification to the user object.
> 
> Reverify option is little bit complicated since we need to define when is user able to reverify - For example there is big difference when the user is kicked from the server and when he leaves on its own. Is only situation when he can reverify when he leaves on its own or do we want to have option to "lift the ban". If the second option is true, then I would suggest leaving it for now and making it separate feature.
> 
> As for the the date of verification. Do you have something in mind how we would use it? Because first of all I would like to keep amount of data we store to minimum (mainly to protect students) and we also have currently quite a few people who are already verfied therefore simplest thing would be to add same date to each one, and I do not find that too useful 🤔 .

In my opinion only re-verification when user leaves server should be available, as currently in this case, the record is still in database. 

Regarding the second request, I see it useful for security reasons when auditing suspicious behaviour... for example cases when there would be some leak from the server, we can filter out users verified in that period and check their behaviour about the leak.
For the users that are already verified, we can use the date of joining the server or just keep the field empty in their case.

**Comment by mimotej** [comment]
Converted back to draft since I am finishing the work on kicked and banned reverification and needs little bit more work.

**Comment by mimotej** [comment]
Now we have also states removed and banned. These states work this way:
removed -> user kicked, left or ban was removed - user can verify again (also new user can verify using proper credentials.
banned -> user is not able to verify again.
verified -> no other user can verify using these credentials

File: src/commands/verify.ts (hunk 1 above)
Comment (mimotej) [reply src/commands/verify.ts:66]: Resolved in latest commit.

File: src/model/schema.prisma (hunk 4 above)
Comment (mimotej) [reply src/model/schema.prisma]: Resolved in latest commit.

**Pull request merged**